### PR TITLE
zenotool now supports fetching metadata about webcache

### PR DESCRIPTION
**Pull request by oxtoacart**

### CODE REVIEW

**Comment by joesis** [comment]
Maybe not new, but CI failed because of `ran too long`.

What is this PR for?

**Comment by oxtoacart** [comment]
This adds the capability to zenotool to list old permalinks from our webcache. I'm using it right now to find old daily bandwidth reports so that we can compare them to our current ones.

File: web/cache.go [hunk 1]
```diff
@@ -113,22 +113,27 @@ func newCache(cacheDir string, ttl time.Duration) (*cache, error) {
...
+func newCacheFromFile(cacheFile string, ttl time.Duration, options *bolt.Options) (*cache, error) {
+	db, err := bolt.Open(cacheFile, 0600, options)
 	if err != nil {
 		return nil, errors.New("Unable to open cache database: %v", err)
 	}
 
-	err = db.Update(func(tx *bolt.Tx) error {
-		_, bucketErr := tx.CreateBucketIfNotExists(cacheBucket)
-		if bucketErr != nil {
+	if !options.ReadOnly {
```
Comment (joesis) [inline web/cache.go]: `options` is nil when calling from `newCache`

Comment (joesis) [reply web/cache.go]: I'm curious why `TestCache` didn't trigger panic.

Comment (joesis) [reply web/cache.go]: It does panic when testing locally. Maybe CI timed out before being able to test this.

Comment (oxtoacart) [reply web/cache.go]: Good catch!

Comment (oxtoacart) [reply web/cache.go]: Fixed

**Comment by oxtoacart** [comment]
It's not good that CI has been failing for a while. I'll fix that as part of this PR.

File: web/toolsupport.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+		return nil, err
+	}
+
+	sort.Sort(byTimestamp(permalinks))
+	return permalinks, nil
+}
+
+func (c *cache) listPermalinks() ([]*PermalinkInfo, error) {
+	var permalinks []*PermalinkInfo
+	err := c.db.View(func(tx *bolt.Tx) error {
```
Comment (joesis) [inline web/toolsupport.go]: Nit: If `listPermalinks` is conceptually considered a feature of the `cache`, the code should be in `cache.go`; if not, maybe make it a reguar function, i.e., `listPermalinks(c *cache)`. Note this is the only case `c` is referenced.

**Comment by joesis** [comment]
Please also apply this patch (update `goexpr` to current master, as 20677f88a18bf93814417b1598cf6b798c0b1056 no longer exists).
```
diff --git a/Gopkg.lock b/Gopkg.lock
index 2dd9398..6a2b9f5 100644
--- a/Gopkg.lock
+++ b/Gopkg.lock
@@ -149,7 +149,7 @@
     "redis",
   ]
   pruneopts = "UT"
-  revision = "20677f88a18bf93814417b1598cf6b798c0b1056"
+  revision = "64ec90b5c995f627388affb6ac64181f1fb7dea0"

 [[projects]]
   branch = "master"
```

File: web/toolsupport.go (hunk 2 above)
Comment (oxtoacart) [reply web/toolsupport.go]: The function needs to be public and cache is not. Conceptually, it only exists to support zenotool, hence it's in `toolsupport.go`.

Comment (joesis) [reply web/toolsupport.go]: Okay

**Comment by joesis** [comment]
LGTM. Compiler complains about two cases of missed parameters when calling `errors.New`, so I fixed them. Merging!

**Comment by oxtoacart** [comment]
Thanks @joesis!

**Pull request merged**